Create stream to stream events from ble scanning functionality merge after #122

torfmaster opened the pull request:
# Summary
In this PR I migrate the ble scanning funtionality from callbacks to `futures::stream::Streams`. 
Only merge after #122 

# Questions/Help needed

 * I decided to make the `BleScanningDriverScanning::stream_values` method to consume `&self` instead of `&mut self`. While I think this is correct it may lead to unexpected behavior when two parallel Tasks consume this stream, i.e. the events are split between the consumers
 * I only implemented infinite streams as the stream of ble events will be either be stopped at runtime (to a point of time predictable at compile time) or continue to produce values forever.
 * I removed the possibilty of registering callbacks to the ble driver directly as for me it seemed to be only feasible for trivial business logic.

Woyten commented:
What is the benefit of implementing `Stream`? Wouldn't a regular `Future<Item=Option<...>>` be equally good? What I don't like about the `Stream` is that you need to implement `poll` manually. Besides that, I am not sure whether `Stream`s will become a part of the `core` library or get deprecated in favor of `async`-`await`.

torfmaster commented:
To be honest I trusted `Stream` beeing in `async_std` a good reason for implementing it in `libtock` as well. I think of a stream of being an async iterator - I wonder how one could implement this differently. What I also like is the combinator functions like filter, map etc. but I have to admit that the `ble_scanning` example doesn't make use of these features. Maybe I should extend the example (e.g. by filtering out unintersting payloads) to make more clear what can be done with streams.

torfmaster commented:
After some experiments I found out that using streams without heap is extremely painful so I changed to a `future` api. Moreover, I rebased the PR on #122 

bors try

torfmaster commented:
bors try

Woyten commented:
Please rebase :wink:

Woyten commented:
@torfmaster Please rebase.

torfmaster commented:
bors try

Woyten commented on examples-alloc/ble_scanning.rs:
```diff
@@ -17,27 +14,19 @@ struct LedCommand {
 #[libtock::main]
 async fn main() -> TockResult<()> {
     let mut drivers = libtock::retrieve_drivers()?;
+    let led_driver = drivers.leds.init_driver()?;
```
Actually, the old name was better.

Woyten commented:
bors r+

Woyten commented:
bors r+
